R/lazyframe-frame.R

```diff
@@ -2313,4 +2313,134 @@ lazyframe__join_asof <- function(
+#'  * `"gzip"`: min-level: 0, max-level: 10.
+#'  * `"brotli"`: min-level: 0, max-level: 11.
+#'  * `"zstd"`: min-level: 1, max-level: 22.
+#' @param statistics Whether statistics should be written to the Parquet
+#' headers. Possible values:
+#' * `TRUE`: enable default set of statistics (default)
+#' * `FALSE`: disable all statistics
+#' * A list created via [parquet_statistics()] to specify which statistics to
+#'   include.
+#' @param row_group_size Size of the row groups in number of rows. If `NULL`
+#' (default), the chunks of the DataFrame are used. Writing in smaller chunks
+#' may reduce memory pressure and improve writing speeds.
+#' @param data_page_size Size of the data page in bytes. If `NULL` (default), it
+#' is set to 1024^2 bytes.
+#' @param maintain_order Maintain the order in which data is processed. Setting
+#' this to `FALSE` will be slightly faster.
+#' @inheritParams lazyframe__collect
+#' @inheritParams pl__scan_parquet
+#'
+#' @return Invisibly returns the input LazyFrame
+#'
+#' @examples
+#' # sink table 'mtcars' from mem to parquet
+#' tmpf <- tempfile()
+#' as_polars_lf(mtcars)$sink_parquet(tmpf)
+#'
+#' # stream a query end-to-end
+#' tmpf2 <- tempfile()
+#' pl$scan_parquet(tmpf)$select(pl$col("cyl") * 2)$sink_parquet(tmpf2)
+#'
+#' # load parquet directly into a DataFrame / memory
+#' pl$scan_parquet(tmpf2)$collect()
+lazyframe__sink_parquet <- function(
+    path,
+    ...,
+    compression = "zstd",
+    compression_level = 3,
+    statistics = TRUE,
+    row_group_size = NULL,
+    data_page_size = NULL,
+    maintain_order = TRUE,
+    type_coercion = TRUE,
+    `_type_check` = TRUE,
+    predicate_pushdown = TRUE,
+    projection_pushdown = TRUE,
+    simplify_expression = TRUE,
+    slice_pushdown = TRUE,
+    collapse_joins = TRUE,
+    no_optimization = FALSE,
+    storage_options = NULL,
+    retries = 2) {
+  wrap({
+    check_dots_empty0(...)
+    compression <- arg_match0(
+      compression,
+      values = c("lz4", "uncompressed", "snappy", "gzip", "lzo", "brotli", "zstd")
+    )
+    lf <- set_sink_optimizations(
+      self,
+      type_coercion = type_coercion,
+      `_type_check` = `_type_check`,
+      predicate_pushdown = predicate_pushdown,
+      projection_pushdown = projection_pushdown,
+      simplify_expression = simplify_expression,
+      slice_pushdown = slice_pushdown,
+      collapse_joins = collapse_joins,
+      no_optimization = no_optimization
+    )
+    if (isTRUE(statistics)) {
+      statistics <- parquet_statistics(
+        min = TRUE,
+        max = TRUE,
+        distinct_count = TRUE,
+        null_count = TRUE
+      )
+    } else if (isFALSE(statistics)) {
+      statistics <- parquet_statistics(
+        min = FALSE,
+        max = FALSE,
+        distinct_count = FALSE,
+        null_count = FALSE
+      )
+    }
+    if (!inherits(statistics, "polars_parquet_statistics")) {
+      abort("`statistics` must be TRUE, FALSE, or a call to `parquet_statistics()`.")
+    }
```

Comment:
I think this is a good compromise: it allows the user to pass a logical or a more detailed list of statistics to write, while ensuring the user doesn't pass unknown stats. Handling this on the R side is also easier than doing it in Rust.

Reply:
Agreed.
However, since `polars_object` is supposed to be an environment class for now (although we may move to S7 later), can you change it to an environment instead of a list?
Otherwise I think the `$` method would cause strange errors.

Reply:
I don't understand, what `polars_object` are you talking about? What should be changed to an environment and would it be a problem to have a list?

Reply:
Sorry, my mistake!